modules/ding_adgangsplatformen/ding_adgangsplatformen.module

```diff
@@ -156,10 +163,56 @@ function ding_adgangsplatformen_openplatform_token_get() {
   return $token;
 }
 
+/**
+ * Provider function.
+ *
+ * Get token for authentication as the library. Should also be used when the
+ * service called don't need access to user information or if no user is logged
+ * in.
+ */
+function ding_adgangsplatformen_openplatform_get_agency_token() {
+  $token = &drupal_static(__FUNCTION__);
+  if (!isset($token)) {
+    $cache = cache_get('ddb_cover_upload_token');
+    if ($cache && $cache->expire > REQUEST_TIME) {
+      $token = $cache->data;
+    }
+    else {
+      $config = ding_adgangsplatformen_get_configuration();
+      $agency_id = variable_get('ting_agency', '');
+
+      $client = new \GuzzleHttp\Client();
+      $response = $client->request('POST', $config['urlAccessToken'], [
+        'form_params' => [
+          'grant_type' => 'password',
+          'username' => '@' . $agency_id,
+          'password' => '@' . $agency_id,
+        ],
+        'auth' => [
+          $config['clientId'],
+          $config['clientSecret'],
+        ],
+      ]);
+
+      $content = $response->getBody()->getContents();
+      $json = json_decode($content, true);
+
+      $token = $json['access_token'];
+
+      // Get token expire minus one day to ensure token do not expire will in
+      // use by services.
+      $expire = REQUEST_TIME + ($json['expires_in'] - 86400);
+      cache_set('ddb_cover_upload_token', $token, 'cache', $expire);
```

Comment:
Change to half not that that is optimal (as it is currently 1 month), but I don't have a better suggestion right now. We could let it expire a request an refresh, but that would require a lot of checks on expire when auth at a service fails and more request between service in auth retries. So it seams to give an better UX to just get a new token before it expires.

Reply:
> Change to half not that that is optimal (as it is currently 1 month), but I don't have a better suggestion right now.

You could say that 1/2 the expiry time means double the amount of requests but we are talking <100 library organisations each making ~24 token requests per year instead of ~12. I do not think that is an issue. If you disagree with 1/2 then at least make it relative: 2/3, 3/4.

> We could let it expire a request a refresh, but that would require a lot of checks on expire when auth at a service fails and more request between service in auth retries. So it seams to give a better UX to just get a new token before it expires.

Totally agree.

Reply:
Lets go with 1/2 expire time